### PR TITLE
New complete signup workflow

### PR DESCRIPTION
**Pull request by enricostano**
Steps to complete a signup:

1. Create a Signup:

        POST /api/v1/signups
        {
          "email":"hola@hola.com"
        }

2. Use the token received via email to build a complete signup form and check if the signup exists:

        GET /api/v1/signups/:token

3. If the signup does exists create a new user putting the token in the headers of the request:

        POST /api/v1/users
        X-katuma-signup-token: token-from-the-email
        {
          "email":"hola@hola.com",
          "username":"hola",
          "password":"iamsosecret",
          "password_confirmation":"iamsosecret"
        }

### CODE REVIEW

File: engines/account/app/controllers/account/api/v1/signups_controller.rb [hunk 1]
```diff
@@ -28,7 +28,7 @@ def show
           if signup
             render status: :ok, json: { email: signup.email }
           else
-            render status: :not_found, json: {}
+            render status: :not_found, nothing: true
           end
```
Comment (andresgutgon) [inline engines/account/app/controllers/account/api/v1/signups_controller.rb]: Show is not needed. We do this on the webapp

File: engines/account/app/controllers/account/api/v1/users_controller.rb [hunk 2]
```diff
@@ -54,22 +73,32 @@ def update
...
           params.permit(:email, :username, :first_name, :last_name, :password, :password_confirmation)
         end
+
+        def load_user
+          @user = ::Account::User.find_by_id(params[:id])
+          render status: :not_found, nothing: true unless @user
+        end
+
+        def load_signup
+          signup_token = request.headers['HTTP_X_KATUMA_SIGNUP_TOKEN']
```
Comment (andresgutgon) [inline engines/account/app/controllers/account/api/v1/users_controller.rb]: This is weird. Why token is not another field of the form?

File: engines/account/app/controllers/account/api/v1/signups_controller.rb (hunk 1 above)
Comment (enricostano) [inline engines/account/app/controllers/account/api/v1/signups_controller.rb]: Yes you do :grin:

File: engines/account/app/controllers/account/api/v1/users_controller.rb (hunk 2 above)
Comment (enricostano) [inline engines/account/app/controllers/account/api/v1/users_controller.rb]: I wanted to keep resource (User in this case) attributes clean in the API and put "authorization" stuff in headers. There is no `token` attribute in `User` model, it's just internal stuff that we use to authorize the use of that endpoint.

File: engines/account/app/services/account/signup_service.rb [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
+    def initialize
     end
 
+    # @param email [String]
     # @return [Account::Signup]
-    def execute
-      if @signup.save
-        SignupMailer.confirm_email(@signup).deliver
+    def create!(email)
+      signup = Signup.new(email: email)
```
Comment (enricostano) [inline engines/account/app/services/account/signup_service.rb]: `Signup.find_by_email(email) || Signup.new(email: email)`

to keep the same token for the same email address

File: engines/account/app/controllers/account/api/v1/users_controller.rb (hunk 2 above)
Comment (andresgutgon) [inline engines/account/app/controllers/account/api/v1/users_controller.rb]: What happen if we want to create users from other places other from confirmation screen? We don't have that token in those cases

Comment (enricostano) [inline engines/account/app/controllers/account/api/v1/users_controller.rb]: I've taking that into account. For instance for invitation workflow, we'll have a similar pattern. We'll need a token confirmation via email and we'll do the same using another custom header.

I'll abstract that "authentication" logic in a module to make the controller clean, don't worry about it.

Comment (andresgutgon) [inline engines/account/app/controllers/account/api/v1/users_controller.rb]: > we'll do the same using another custom header.

Buff. In both cases we just need to send a form with email and token fields. And in both cases API endpoind need to check those fields. I think doing this in headers is overengeniering in frontend and backend

Comment (enricostano) [inline engines/account/app/controllers/account/api/v1/users_controller.rb]: talked offline

**Pull request merged**